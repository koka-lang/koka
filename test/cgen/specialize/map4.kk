
fun map-poly(xs : list<a>, f : a -> e b) : e list<b> {
  match (xs) {
    Nil -> Nil
    Cons(x, xx) -> Cons(x.f, xx.map-poly(f))    
  }
}

public fun test(xs : list<a>, y : int, g : a -> int) {
  xs.map-poly(fn(x){ g(x) + y })   // free variable  
}

public fun main() {
  list(1,10).test(5, fn(x){ x*2 }).sum.println
}
