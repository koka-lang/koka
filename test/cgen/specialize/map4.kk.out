160
 
cgen/specialize/map4/.lift000-test: forall<a> (g : (a) -> int, y : int, xs0 : list<a>) -> list<int>
cgen/specialize/map4/.lift000-main: (xs : list<int>) -> list<int>
cgen/specialize/map4/.mlift000-map-poly: forall<a,e> (a, list<a>) -> e list<a>
cgen/specialize/map4/.mlift000-map-poly: forall<a,b,e> (f : (a) -> e b, xx : list<a>, b) -> e list<b>
cgen/specialize/map4/main: () -> console ()
cgen/specialize/map4/map-poly: forall<a,b,e> (xs : list<a>, f : (a) -> e b) -> e list<b>
cgen/specialize/map4/test: forall<a> (xs : list<a>, y : int, g : (a) -> int) -> list<int>